lib/Containers/MerkleTree.cpp

```diff
@@ -455,46 +537,66 @@ void MerkleTree<Hasher, BranchingBits, LockStripes>::remove(std::vector<std::uin
     throw std::out_of_range("Cannot remove, key out of current range.");
   }
 
-  modify(sortedKeys, false);
+  modify(sortedKeys, /*isInsert*/ false);
 }
 
-template <typename Hasher, std::uint64_t const BranchingBits, std::uint64_t const LockStripes>
-void MerkleTree<Hasher, BranchingBits, LockStripes>::clear() {
+template <typename Hasher, std::uint64_t const BranchingBits>
+void MerkleTree<Hasher, BranchingBits>::clear() {
   std::unique_lock<std::shared_mutex> guard(_bufferLock);
+
   std::uint64_t const last = nodeCountUpToDepth(meta().maxDepth);
   for (std::uint64_t i = 0; i < last; ++i) {
     new (&node(i)) Node{0, 0};
   }
 }
 
-template <typename Hasher, std::uint64_t const BranchingBits, std::uint64_t const LockStripes>
-std::unique_ptr<MerkleTree<Hasher, BranchingBits, LockStripes>>
-MerkleTree<Hasher, BranchingBits, LockStripes>::clone() {
+template <typename Hasher, std::uint64_t const BranchingBits>
+void MerkleTree<Hasher, BranchingBits>::checkConsistency() const {
+  std::shared_lock<std::shared_mutex> guard(_bufferLock);
+
+  checkInternalConsistency();
+}
+
+template <typename Hasher, std::uint64_t const BranchingBits>
+std::unique_ptr<MerkleTree<Hasher, BranchingBits>> MerkleTree<Hasher, BranchingBits>::clone() {
+  // acquire the read-lock here to protect ourselves from concurrent modifications
   std::shared_lock<std::shared_mutex> guard(_bufferLock);
-  return std::unique_ptr<MerkleTree<Hasher, BranchingBits, LockStripes>>(
-      new MerkleTree<Hasher, BranchingBits, LockStripes>(*this));
+
+  return std::unique_ptr<MerkleTree<Hasher, BranchingBits>>(
+      new MerkleTree<Hasher, BranchingBits>(*this));
 }
 
-template <typename Hasher, std::uint64_t const BranchingBits, std::uint64_t const LockStripes>
-std::unique_ptr<MerkleTree<Hasher, BranchingBits, LockStripes>>
-MerkleTree<Hasher, BranchingBits, LockStripes>::cloneWithDepth(std::uint64_t newDepth) const {
+template <typename Hasher, std::uint64_t const BranchingBits>
+std::unique_ptr<MerkleTree<Hasher, BranchingBits>>
+MerkleTree<Hasher, BranchingBits>::cloneWithDepth(std::uint64_t newDepth) const {
+  // arangod does not call this method!
```

Comment:
```suggestion
  // arangod does not call this method! That is, because it is rather dangerous:
  // To grow the depth, it has to modify `rangeMax` by multiplying `rangeMax-rangeMin` by
  // the `branchingFactor`. This can quickly lead to integer overflow! Do not use without
  // knowing exactly what you are doing!
```